Normalize ident

crlf0710 opened the pull request:
Perform unicode normalization on identifiers. Resolving the first bullet point in #55467.

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

crlf0710 commented:
r? @estebank
And i think will hurt performance a little. Maybe we should measure it before it lands.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14232&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14232/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14232/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-23T14:44:27.8437353Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-23T14:44:27.8453281Z ##[command]git config gc.auto 0
2019-11-23T14:44:27.8499331Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-23T14:44:27.8540359Z ##[command]git config --get-all http.proxy
2019-11-23T14:44:27.8663581Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66670/merge:refs/remotes/pull/66670/merge
---
2019-11-23T15:33:31.2561630Z .................................................................................................... 1600/9281
2019-11-23T15:33:35.4769065Z .................................................................................................... 1700/9281
2019-11-23T15:33:46.6108886Z ...........................i........................................................................ 1800/9281
2019-11-23T15:33:52.2225717Z .................................................................................................... 1900/9281
2019-11-23T15:34:04.2520439Z ............iiiii................................................................................... 2000/9281
2019-11-23T15:34:12.2595071Z .................................................................................................... 2200/9281
2019-11-23T15:34:14.7292704Z .................................................................................................... 2300/9281
2019-11-23T15:34:19.2892354Z .................................................................................................... 2400/9281
2019-11-23T15:34:36.5565960Z .................................................................................................... 2500/9281
---
2019-11-23T15:36:58.2344464Z ............i...............i....................................................................... 4800/9281
2019-11-23T15:37:06.6957700Z .................................................................................................... 4900/9281
2019-11-23T15:37:11.5742193Z .................................................................................................... 5000/9281
2019-11-23T15:37:19.7502276Z .................................................................................................... 5100/9281
2019-11-23T15:37:25.2078253Z .................ii.ii...........i.................................................................. 5200/9281
2019-11-23T15:37:33.5455009Z .................................................................................................... 5400/9281
2019-11-23T15:37:43.2590696Z ...................................................................................................i 5500/9281
2019-11-23T15:37:50.5966007Z .................................................................................................... 5600/9281
2019-11-23T15:37:55.9819282Z .................................................................................................... 5700/9281
2019-11-23T15:37:55.9819282Z .................................................................................................... 5700/9281
2019-11-23T15:38:05.3697115Z .....................................................................................ii...i..ii..... 5800/9281
2019-11-23T15:38:25.1355922Z .................................................................................................... 6000/9281
2019-11-23T15:38:31.8522220Z .................................................................................................... 6100/9281
2019-11-23T15:38:35.6726851Z .................................................................................................... 6200/9281
2019-11-23T15:38:35.6726851Z .................................................................................................... 6200/9281
2019-11-23T15:38:46.8930340Z ........i..ii....................................................................................... 6300/9281
2019-11-23T15:39:02.3584974Z ............................................................................i....................... 6500/9281
2019-11-23T15:39:04.4406845Z .................................................................................................... 6600/9281
2019-11-23T15:39:06.3849978Z ...................................................................i................................ 6700/9281
2019-11-23T15:39:08.8141806Z .................................................................................................... 6800/9281
---
2019-11-23T15:43:05.8447433Z ---- [ui] ui/codemap_tests/unicode_2.rs stdout ----
2019-11-23T15:43:05.8447480Z diff of stderr:
2019-11-23T15:43:05.8447527Z 
2019-11-23T15:43:05.8447558Z 14    |
2019-11-23T15:43:05.8447593Z 15    = help: valid widths are 8, 16, 32, 64 and 128
2019-11-23T15:43:05.8447831Z - error[E0425]: cannot find value `a̐é` in this scope
2019-11-23T15:43:05.8448008Z + error[E0425]: cannot find value `a̐é` in this scope
2019-11-23T15:43:05.8448184Z 18   --> $DIR/unicode_2.rs:6:13
2019-11-23T15:43:05.8448235Z 19    |
2019-11-23T15:43:05.8448235Z 19    |
2019-11-23T15:43:05.8448382Z 20 LL |     let _ = a̐é;
2019-11-23T15:43:05.8448404Z 
2019-11-23T15:43:05.8448422Z 
2019-11-23T15:43:05.8448471Z The actual stderr differed from the expected stderr.
2019-11-23T15:43:05.8448700Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codemap_tests/unicode_2/unicode_2.stderr
2019-11-23T15:43:05.8448887Z To update references, rerun the tests and pass the `--bless` flag
2019-11-23T15:43:05.8449099Z To only update this specific test, also pass `--test-args codemap_tests/unicode_2.rs`
2019-11-23T15:43:05.8449157Z error: 1 errors occurred comparing output.
2019-11-23T15:43:05.8449191Z status: exit code: 1
2019-11-23T15:43:05.8449191Z status: exit code: 1
2019-11-23T15:43:05.8449899Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/codemap_tests/unicode_2.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codemap_tests/unicode_2" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codemap_tests/unicode_2/auxiliary" "-A" "unused"
2019-11-23T15:43:05.8450207Z ------------------------------------------
2019-11-23T15:43:05.8450233Z 
2019-11-23T15:43:05.8450392Z ------------------------------------------
2019-11-23T15:43:05.8450441Z stderr:
2019-11-23T15:43:05.8450441Z stderr:
2019-11-23T15:43:05.8450782Z ------------------------------------------
2019-11-23T15:43:05.8451221Z error: invalid width `7` for integer literal
2019-11-23T15:43:05.8451586Z    |
2019-11-23T15:43:05.8451586Z    |
2019-11-23T15:43:05.8451829Z LL |     let _ = ("a̐éö̲", 0u7); //~ ERROR invalid width
2019-11-23T15:43:05.8451955Z    |
2019-11-23T15:43:05.8451955Z    |
2019-11-23T15:43:05.8452001Z    = help: valid widths are 8, 16, 32, 64 and 128
2019-11-23T15:43:05.8452034Z 
2019-11-23T15:43:05.8452098Z error: invalid width `42` for integer literal
2019-11-23T15:43:05.8452425Z    |
2019-11-23T15:43:05.8452425Z    |
2019-11-23T15:43:05.8452713Z LL |     let _ = ("아あ", 1i42); //~ ERROR invalid width
2019-11-23T15:43:05.8452809Z    |
2019-11-23T15:43:05.8452809Z    |
2019-11-23T15:43:05.8452876Z    = help: valid widths are 8, 16, 32, 64 and 128
2019-11-23T15:43:05.8453165Z error[E0425]: cannot find value `a̐é` in this scope
2019-11-23T15:43:05.8453424Z   --> /checkout/src/test/ui/codemap_tests/unicode_2.rs:6:13
2019-11-23T15:43:05.8453489Z    |
2019-11-23T15:43:05.8453725Z LL |     let _ = a̐é; //~ ERROR cannot find
---
2019-11-23T15:43:05.8479568Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-23T15:43:05.8479634Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-23T15:43:05.8491102Z 
2019-11-23T15:43:05.8491212Z 
2019-11-23T15:43:05.8494428Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-23T15:43:05.8494915Z 
2019-11-23T15:43:05.8495070Z 
2019-11-23T15:43:05.8504576Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-23T15:43:05.8504636Z Build completed unsuccessfully in 0:53:08
2019-11-23T15:43:05.8504636Z Build completed unsuccessfully in 0:53:08
2019-11-23T15:43:05.8548820Z == clock drift check ==
2019-11-23T15:43:05.8560947Z   local time: Sat Nov 23 15:43:05 UTC 2019
2019-11-23T15:43:06.3766966Z   network time: Sat, 23 Nov 2019 15:43:06 GMT
2019-11-23T15:43:06.3770366Z == end clock drift check ==
2019-11-23T15:43:07.2796943Z 
2019-11-23T15:43:07.2877973Z ##[error]Bash exited with code '1'.
2019-11-23T15:43:07.2907889Z ##[section]Starting: Checkout
2019-11-23T15:43:07.2909365Z ==============================================================================
2019-11-23T15:43:07.2909406Z Task         : Get sources
2019-11-23T15:43:07.2909441Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented on src/test/ui/rfc-2457/idents-normalized.rs:
```diff
@@ -0,0 +1,8 @@
+// run-pass
```
```suggestion
// check-pass
```

crlf0710 replied:
Done.

Centril commented:
cc @Manishearth

Centril commented:
cc @petrochenkov re. Symbols

crlf0710 commented:
> hopefully there are no other places where we create Symbols?

At least there's one in the `proc_macro` crate. I looked at the code but not sure how should i change it. I know nothing about the "bridge" thing. I'm leaving that into a further PR.

estebank commented:
@Manishearth shouldn't we have a warning when encountering non-NFC chars? Shouldn't we, for the sake of third party tools, attempt to "soft-enforce" byte comparable tokens? These questions might are possibly unrelated to this PR in particular and I should have brought up this point in the mega-thread.

estebank commented:
The PR LGTM, but I can foresee issues cropping up when involving macros.

Manishearth commented:
@estebank no, non-NFC chars appear for all kinds of reasons

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 0e3036b5da9f64c27878150b84a22903663f598f with merge ce0c4f69329436f488883ad75290b9e38d921fed...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0e3036b5da9f64c27878150b84a22903663f598f","merge_sha":"ce0c4f69329436f488883ad75290b9e38d921fed"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14465)
Build commit: ce0c4f69329436f488883ad75290b9e38d921fed (`ce0c4f69329436f488883ad75290b9e38d921fed`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14465"},"merge_sha":"ce0c4f69329436f488883ad75290b9e38d921fed"} -->

rust-timer commented:
Queued ce0c4f69329436f488883ad75290b9e38d921fed with parent a44774c3a9739b2eea8923e09d67b14312c78ef3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a44774c3a9739b2eea8923e09d67b14312c78ef3&end=ce0c4f69329436f488883ad75290b9e38d921fed).

crlf0710 commented:
Mmm, is the timer task still queued?

crlf0710 commented:
@bors try @rust-timer queue

bors commented:
@crlf0710: :key: Insufficient privileges: not in try users

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.

est31 commented on src/test/ui/rfc-2457/idents-normalized.rs:
```diff
@@ -0,0 +1,8 @@
+struct Résumé; // ['LATIN SMALL LETTER E WITH ACUTE']
+
+fn main() {
+    let _ = Résumé; // ['LATIN SMALL LETTER E', 'COMBINING ACUTE ACCENT']
```
Some folks in the hypothetical future might run rustfmt on all tests. And there were talks about rustfmt doing normalization. I wonder whether there should be some verification that these literals indeed contain those characters otherwise the test might be rendered ineffective in the future.

Dylan-DPC commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout normalize_ident` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self normalize_ident --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/librustc_parse/lexer/mod.rs
Auto-merging src/librustc_parse/Cargo.toml
CONFLICT (content): Merge conflict in src/librustc_parse/Cargo.toml
Auto-merging Cargo.lock
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

Dylan-DPC commented:
@crlf0710 can you rebase this and resolve the merge conflict so we can perf this?

bors commented:
:umbrella: The latest upstream changes (presumably #67220) made this pull request unmergeable. Please resolve the merge conflicts.

crlf0710 commented:
Maybe we can try perf it again? @Dylan-DPC

Dylan-DPC commented:
@rust-timer queue

rust-timer commented:
Awaiting bors try build completion

crlf0710 commented:
@Dylan-DPC I'm not familiar with the perf tool... What's the current status?

estebank commented:
We're in the beta cutoff, so that gives us 6 weeks to catch any real regressions, which I don't anticipate to begin with.

@bors r+

This PR will be merged once the beta has been cut for release.

bors commented:
:pushpin: Commit 4e0a4e000ac6947c6356416429d559fb800f84ba has been approved by `estebank`

<!-- @bors r=estebank 4e0a4e000ac6947c6356416429d559fb800f84ba -->
<!-- homu: {"type":"Approved","sha":"4e0a4e000ac6947c6356416429d559fb800f84ba","approver":"estebank"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 100, this pull request will be tested once the tree is reopened

Mark-Simulacrum commented:
(Just to make sure everyone is on the same page here: beta has already been cut)

Mark-Simulacrum commented:
Also going to raise this above tree closure priority... @bors p=100

estebank commented:
When is the tree reopened?

bors commented:
:hourglass: Testing commit 4e0a4e000ac6947c6356416429d559fb800f84ba with merge d57667544396e25e33a5df417a75f861fa264e13...
<!-- homu: {"type":"BuildStarted","head_sha":"4e0a4e000ac6947c6356416429d559fb800f84ba","merge_sha":"d57667544396e25e33a5df417a75f861fa264e13"} -->

rust-highfive commented:
The job `i686-apple` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=16219&view=logs&jobId=8ac55db3-a6cf-5386-3be9-05e8c37150db) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16219/logs/1333), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16219/logs/1333)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-19T03:37:39.8421220Z Done!
2019-12-19T03:37:39.8421780Z some tidy checks failed
2019-12-19T03:37:39.8425320Z 
2019-12-19T03:37:39.8425870Z 
2019-12-19T03:37:39.8427740Z command did not execute successfully: "/Users/runner/runners/2.163.1/work/1/s/build/i686-apple-darwin/stage0-tools-bin/tidy" "/Users/runner/runners/2.163.1/work/1/s/build/i686-apple-darwin/stage0-tools-bin/tidy" "/Users/runner/runners/2.163.1/work/1/s/src" "/Users/runner/runners/2.163.1/work/1/s/build/i686-apple-darwin/stage0/bin/cargo" "--no-vendor"
2019-12-19T03:37:39.8428950Z 
2019-12-19T03:37:39.8429460Z 
2019-12-19T03:37:39.8438450Z failed to run: /Users/runner/runners/2.163.1/work/1/s/build/bootstrap/debug/bootstrap test
2019-12-19T03:37:39.8439130Z Build completed unsuccessfully in 0:01:35
2019-12-19T03:37:39.8439130Z Build completed unsuccessfully in 0:01:35
2019-12-19T03:37:39.8510280Z == clock drift check ==
2019-12-19T03:37:39.8576320Z   local time: Thu Dec 19 03:37:39 UTC 2019
2019-12-19T03:37:39.8859760Z   network time: Thu, 19 Dec 2019 03:37:39 GMT
2019-12-19T03:37:39.8860170Z == end clock drift check ==
2019-12-19T03:37:39.8924790Z 
2019-12-19T03:37:39.9047890Z ##[error]Bash exited with code '1'.
2019-12-19T03:37:39.9101830Z ##[section]Starting: Checkout
2019-12-19T03:37:39.9104670Z ==============================================================================
2019-12-19T03:37:39.9104780Z Task         : Get sources
2019-12-19T03:37:39.9104860Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=16219)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=16219","builder_name":"checks-azure"} -->

crlf0710 commented:
```
2019-12-19T03:37:39.1564050Z Dependencies not on the whitelist:
2019-12-19T03:37:39.1570810Z * unicode-normalization
```
Rebased and added `unicode-normalization` to the whitelist. @estebank

Mark-Simulacrum commented:
@bors r=estebank p=0

bors commented:
:pushpin: Commit 49f3bc93ec3b434ab52deaed9678749a6170372f has been approved by `estebank`

<!-- @bors r=estebank 49f3bc93ec3b434ab52deaed9678749a6170372f -->
<!-- homu: {"type":"Approved","sha":"49f3bc93ec3b434ab52deaed9678749a6170372f","approver":"estebank"} -->

estebank commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

bors commented:
:pushpin: Commit 49f3bc93ec3b434ab52deaed9678749a6170372f has been approved by `estebank`

<!-- @bors r=estebank 49f3bc93ec3b434ab52deaed9678749a6170372f -->
<!-- homu: {"type":"Approved","sha":"49f3bc93ec3b434ab52deaed9678749a6170372f","approver":"estebank"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 100, this pull request will be tested once the tree is reopened

Mark-Simulacrum commented:
I believe this caused the rollup failure here: https://github.com/rust-lang/rust/pull/67435#issuecomment-567708950

You can likely fix it by adding smallvec at version 1.0 (not 0.6!) to https://github.com/rust-lang/rust/blob/master/src/tools/rustc-workspace-hack/Cargo.toml#L65 and use Cargo's package renaming to make sure those two don't conflict as keys in the dependencies section.

@bors r-

crlf0710 commented:
Rebased and add dependency according to @Mark-Simulacrum 's instruction. @estebank

Dylan-DPC commented:
@bors r+

bors commented:
:pushpin: Commit 27e7a1baedbcc5ddaf44f930860828dae99a7ebf has been approved by `Dylan-DPC`

<!-- @bors r=Dylan-DPC 27e7a1baedbcc5ddaf44f930860828dae99a7ebf -->
<!-- homu: {"type":"Approved","sha":"27e7a1baedbcc5ddaf44f930860828dae99a7ebf","approver":"Dylan-DPC"} -->

bors commented:
:hourglass: Testing commit 27e7a1baedbcc5ddaf44f930860828dae99a7ebf with merge acb6690e1d58fc5f262ada5b5030fe73e601f1e8...
<!-- homu: {"type":"BuildStarted","head_sha":"27e7a1baedbcc5ddaf44f930860828dae99a7ebf","merge_sha":"acb6690e1d58fc5f262ada5b5030fe73e601f1e8"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=16826)
Approved by: Dylan-DPC
Pushing acb6690e1d58fc5f262ada5b5030fe73e601f1e8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Dylan-DPC","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=16826"},"merge_sha":"acb6690e1d58fc5f262ada5b5030fe73e601f1e8"} -->

petrochenkov commented:
I have two questions / concerns:
- Is it possible to normalize a non-ASCII identifier into an ASCII identifier?
- Even without that non-ASCII identifiers are gated at the wrong point (AST after macro expansion) and not entirely correctly (`visit_name` instead of identifier tokens), rather than at the correct point (during lexing).
Is there any desire to fix the gating (do it in lexer before the normalization) before the feature is stabilized?

crlf0710 commented:
Following the RFC, since we're using NFC normalization, I think the answer to the first question is "no it's not possible". NFC normalization is basically reordering and regrouping, so nothing really disappears.

crlf0710 commented:
For the second question, i think that's a question for a potential "shepherd" for RFC 2457, currently we don't have one, but it would be best if there is one. cc #55467

In my personal opinion though, i think it's not too wrong to gate at the AST after macro expansion, since macros and procedure macros can generate new identifiers, and need to be gated too. I basically know nothing about the `(visit_name instead of identifier tokens)` part.

crlf0710 commented:
And by the way i believe there're some remaining work on performing this normalization to the user data sent to `proc_macro` crate.  the RFC mentioned that we need to make sure: 

> Every API accepting raw identifiers (such as proc_macro::Ident::new normalizes them to NFC and APIs returning them as strings (like proc_macro::Ident::to_string) return the normalized form.

I think I need some mentor here. @petrochenkov could you give some instructions when you have time? I tried to read the code in `proc_macro` crate, but got total lost in the `bridge`-related thing.  It's also concerning if we need to add the `unicode-normalization` dependency to the `proc_macro` crate.

petrochenkov commented:
@crlf0710 
The "cannot be a raw identifier" error is an example of catching something at lexer level.

It's emitted twice in `librustc_parse\lexer\mod.rs` and in `libsyntax_expand\proc_macro_server.rs` and those are the only places where an identifier token can be created from user input.

So, the normalization (and also gating, arguably) needs to happen in the same places.
(It will catch anything produced by macros as well.)

crlf0710 commented:
@petrochenkov do you mean https://github.com/rust-lang/rust/blob/master/src/libsyntax_expand/proc_macro_server.rs#L330 ? 

I'm a little confused. After normalization, if the identifier changes, should i intern, get a new `Symbol` and use it in the result identifier?  What should i do there with the existing `Symbol`, leaving it alone?

petrochenkov commented:
@crlf0710 
Yes.

crlf0710 commented:
Great, let me create a branch and PR it.

petrochenkov commented:
By the way, the RFC says the normalization is performed during parsing rather than lexing.
That means proc macros receive the original unnormalized tokens and can perform their own normalization (same as with escaping in literals).
I'm not sure that's the intent of the RFC.

On the other hand, if the normalization is performed during lexing, that would be the first case in which we do not preserve original tokens by design (except for a big hack with doc comments in macros).
